File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) {
            return null;
        }
        
        Map<Node, Node> map = new HashMap<>();
        map.put(head, new Node(head.val));
        Node oriCur = head;
        Node cur = map.get(head);

        while (oriCur != null) {
            if (oriCur.next != null) {
                if (!map.containsKey(oriCur.next)) {
                    map.put(oriCur.next, new Node(oriCur.next.val));
                }
                cur.next = map.get(oriCur.next);
            }
            if (oriCur.random != null) {
                if (!map.containsKey(oriCur.random)) {
                    map.put(oriCur.random, new Node(oriCur.random.val));
                }
                cur.random = map.get(oriCur.random);
            }

            oriCur = oriCur.next;
            cur = cur.next;
        }

        return map.get(head);
    }
}

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if (head == nullptr) {
            return head;
        }
        
        unordered_map<Node*, Node*> map;
        Node* copyHead = new Node(head->val);
        map[head] = copyHead;

        Node* cur = copyHead;
        while (head != nullptr) {
            if (head->next != nullptr) {
                if (map.find(head->next) == map.end()) {
                    Node* newNode = new Node(head->next->val);
                    map[head->next] = newNode;
                }
                cur->next = map[head->next];
            }
            if (head->random != nullptr) {
                if (map.find(head->random) == map.end()) {
                    Node* newNode = new Node(head->random->val);
                    map[head->random] = newNode;
                }
                cur->random = map[head->random];
            }

            cur = cur->next;
            head = head->next;
        }

        return copyHead;
    }
};